### PR TITLE
[New Model]Donut model

### PR DESCRIPTION
**Pull request by princepride**
## Purpose

FIX #18850.
This PR adds support for Donut-like models, resolving issue #18850.

It implements the Donut model and the structurally similar Dolphin model. Since the Donut model uses a Swin Transformer as its vision backbone, this PR also includes the implementation for the Swin model.

### Donut Model Test
**Script**

```bash
python examples/offline_inference/encoder_decoder_multimodal.py -m donut
```
**Result**

image

### Dolphin Model Tests
**Script**

```bash
python examples/offline_inference/dolphin.py
```
**Result**

Screenshot 2025-08-22 at 11 03 30 AM

### CODE REVIEW

**Comment by princepride** [comment]
@DarkLight1337 @Isotr0py Could you review it? I messed up the previous PR (https://github.com/vllm-project/vllm/pull/23187) during a rebase, so I've created a new one. Thank you.

File: docs/models/supported_models.md [hunk 1]
```diff
@@ -815,4 +816,4 @@ We have the following levels of testing for models:
...
 2. **Output Sensibility**: We check if the output of the model is sensible and coherent, by measuring the perplexity of the output and checking for any obvious errors. This is a less stringent test.
 3. **Runtime Functionality**: We check if the model can be loaded and run without errors. This is the least stringent test. Please refer to [functionality tests](gh-dir:tests) and [examples](gh-dir:examples) for the models that have passed this test.
-4. **Community Feedback**: We rely on the community to provide feedback on the models. If a model is broken or not working as expected, we encourage users to raise issues to report it or open pull requests to fix it. The rest of the models fall under this category.
+4. **Community Feedback**: We rely on the community to provide feedback on the models. If a model is broken or not working as expected, we encourage users to raise issues to report it or open pull requests to fix it. The rest of the models fall under this category.
```
Comment (DarkLight1337) [inline docs/models/supported_models.md]: Please revert this change

File: tests/models/registry.py [hunk 2]
```diff
@@ -517,6 +517,8 @@ def check_available_online(
                                                          trust_remote_code=True),  # noqa: E501
     "MllamaForConditionalGeneration": _HfExamplesInfo("meta-llama/Llama-3.2-11B-Vision-Instruct"),  # noqa: E501
     "WhisperForConditionalGeneration": _HfExamplesInfo("openai/whisper-large-v3"),  # noqa: E501
+    "DonutForConditionalGeneration": _HfExamplesInfo("naver-clova-ix/donut-base-finetuned-docvqa",  # noqa: E501
```
Comment (DarkLight1337) [inline tests/models/registry.py]: Add Dolphin as an `extra` model

File: tests/models/registry.py [hunk 3]
```diff
@@ -606,4 +608,4 @@ def find_hf_info(self, model_id: str) -> _HfExamplesInfo:
 
 
 HF_EXAMPLE_MODELS = HfExampleModels(_EXAMPLE_MODELS)
-AUTO_EXAMPLE_MODELS = HfExampleModels(_AUTOMATIC_CONVERTED_MODELS)
+AUTO_EXAMPLE_MODELS = HfExampleModels(_AUTOMATIC_CONVERTED_MODELS)
```
Comment (DarkLight1337) [inline tests/models/registry.py]: Revert this change

File: vllm/model_executor/models/registry.py [hunk 4]
```diff
@@ -856,4 +857,4 @@ def _run() -> None:
 
 
 if __name__ == "__main__":
-    _run()
+    _run()
```
Comment (DarkLight1337) [inline vllm/model_executor/models/registry.py]: Same

File: vllm/model_executor/models/swin.py [hunk 5]
```diff
@@ -0,0 +1,909 @@
...
+
+from vllm.model_executor.layers.activation import get_act_fn
+from vllm.model_executor.layers.linear import (ColumnParallelLinear,
+                                               QKVParallelLinear,
+                                               RowParallelLinear)
+from vllm.model_executor.layers.quantization import QuantizationConfig
+from vllm.model_executor.model_loader.weight_utils import default_weight_loader
+
+
+# Copied from transformers.models.swin.modeling_swin.window_partition
```
Comment (DarkLight1337) [inline vllm/model_executor/models/swin.py]: We can import these directly from transformers library

File: vllm/model_executor/models/donut.py [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+            encoder_positions
+                torch.Tensor of *encoder* position indices
+        Returns:
+            Output torch.Tensor
+        """
+
+        inputs_embeds = None
+        if encoder_input_ids.numel() > 0:
+            vision_embeddings = self.get_multimodal_embeddings(**kwargs)
+            inputs_embeds = _flatten_embeddings(vision_embeddings)
```
Comment (Isotr0py) [inline vllm/model_executor/models/donut.py]:
```suggestion
        if encoder_input_ids.numel() > 0:
            vision_embeddings = self.get_multimodal_embeddings(**kwargs)
            inputs_embeds = self.get_input_embeddings(encoder_input_ids,
                                                      vision_embeddings)
```
Let's move `_flatten_embeddings(vision_embeddings)` into `get_input_embeddings`

File: vllm/model_executor/models/swin.py [hunk 7]
```diff
@@ -0,0 +1,909 @@
...
+        prefix: str = "",
+    ) -> None:
+        super().__init__()
+        self.dense = RowParallelLinear(
+            input_size=dim,
+            output_size=dim,
+            quant_config=quant_config,
+            prefix=f"{prefix}.dense",
+        )
+        self.dropout = nn.Dropout(config.attention_probs_dropout_prob)
```
Comment (Isotr0py) [inline vllm/model_executor/models/swin.py]:
```suggestion
```

File: vllm/model_executor/models/swin.py [hunk 8]
```diff
@@ -0,0 +1,909 @@
...
+                 config: SwinConfig,
+                 dim: int,
+                 quant_config: Optional[QuantizationConfig] = None,
+                 prefix: str = "") -> None:
+        super().__init__()
+        self.dense = RowParallelLinear(int(config.mlp_ratio * dim),
+                                       dim,
+                                       quant_config=quant_config,
+                                       prefix=f"{prefix}.dense")
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
```
Comment (Isotr0py) [inline vllm/model_executor/models/swin.py]:
```suggestion
```

File: vllm/model_executor/models/swin.py [hunk 9]
```diff
@@ -0,0 +1,909 @@
...
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        hidden_states, _ = self.dense(hidden_states)
+        hidden_states = self.dropout(hidden_states)
+        return hidden_states
+
+
+class SwinLayer(nn.Module):
```
Comment (Isotr0py) [inline vllm/model_executor/models/swin.py]: I think we can inherit Swin components from Transformers' implementation: https://github.com/huggingface/transformers/blob/0f9c9088d0260565cd89af2925836ea2f8bfe5f3/src/transformers/models/swin/modeling_swin.py#L573-L585

File: examples/offline_inference/dolphin.py
Comment (Isotr0py) [inline examples/offline_inference/dolphin.py:1]: Let's add examples in `examples/offline_inference/encoder_decoder_multimodal.py` and `examples/offline_inference/vision_language.py` instead of adding new files in `examples/offline_inference/`.

File: vllm/model_executor/models/swin.py (hunk 7 above)
Comment (princepride) [reply vllm/model_executor/models/swin.py]: Why we don't need self.dropout 🤔?

Comment (DarkLight1337) [reply vllm/model_executor/models/swin.py]: Dropout is normally disabled during inference. Can you find a HF Hub config that has dropout enabled?

Comment (princepride) [reply vllm/model_executor/models/swin.py]: https://huggingface.co/ByteDance/Dolphin/blob/main/config.json#L120
https://huggingface.co/naver-clova-ix/donut-base/blob/main/config.json#L116
I checked the config files of several Donut models and indeed the value of hidden_dropout_prob is 0.0, but it may not be 0.0 for some models? I'm not sure.

Comment (DarkLight1337) [reply vllm/model_executor/models/swin.py]: We can assume it to be 0.0. In general vLLM assumes no dropout

Comment (princepride) [reply vllm/model_executor/models/swin.py]: 👌

File: examples/offline_inference/dolphin.py
Comment (princepride) [reply examples/offline_inference/dolphin.py:1]: I was looking at the images in the [S3 bucket](https://vllm-public-assets.s3.us-west-2.amazonaws.com/) and it seems there aren't any suitable for OCR tasks. This is particularly true for the Dolphin model, whose OCR task is similar to executing a workflow. Different prompts will determine whether to segment or parse the document. At the same time, depending on the parsing tags, it will decide whether to parse text or icons. That's why I've added two example files.

File: vllm/model_executor/models/swin.py [hunk 10]
```diff
@@ -0,0 +1,498 @@
...
+        batch_size, dim, num_channels = hidden_states.shape
+
+        qkv_output, _ = self.qkv(hidden_states)
+        query_layer, key_layer, value_layer = qkv_output.chunk(3, dim=-1)
+
+        key_layer = self.transpose_for_scores(key_layer)
+        value_layer = self.transpose_for_scores(value_layer)
+        query_layer = self.transpose_for_scores(query_layer)
+
+        if self.fused_attn:
```
Comment (DarkLight1337) [inline vllm/model_executor/models/swin.py]: `self.fused_attn` seems to be hardcoded to `True`, so we can remove the code in the `else` branch

File: examples/offline_inference/dolphin.py
Comment (princepride) [reply examples/offline_inference/dolphin.py:1]: I used `fetch_image` to load the image and moved `donut` to `encoder_decoder_multimodal.py`

Comment (princepride) [reply examples/offline_inference/dolphin.py:1]: I also removed the `--task` in `dolphin`

**Comment by princepride** [comment]
@Isotr0py @DarkLight1337 I adjust the example code and update the task plan, please review it. thank you.

**Review by DarkLight1337 — APPROVED** [review]
Thanks for the hard work, I think this looks good now.

@Isotr0py can you confirm?

File: vllm/model_executor/models/donut.py [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+        return {
+            "image":
+            self._get_dummy_images(width=target_width,
+                                   height=target_height,
+                                   num_images=num_images)
+        }
+
+
+class DonutMultiModalProcessor(EncDecMultiModalProcessor[DonutProcessingInfo]):
+
```
Comment (Isotr0py) [inline vllm/model_executor/models/donut.py:180]: Let's add processor test in `tests/models/multimodal/processing/test_common.py` to include this model.

**Review by Isotr0py — APPROVED** [review]
```
Adding requests: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.97s/it]
Processed prompts: 100%|█████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.57it/s, est. speed input: 20.46 toks/s, output: 17.31 toks/s]
Decoder prompt: '<s_docvqa><s_question>What time is the coffee break?</s_question><s_answer>', Generated text: ' 11-14 to 11:39 a.m.</s_answer>'
Duration: 3.6066391468048096
RPS: 0.2772664409429258
```

The example outputs look reasonable, LGTM!

**Comment by princepride** [comment]
I noticed there was a unit test error caused by my code, and I will fix it.

**Comment by princepride** [comment]
@DarkLight1337 I think it's ready to merge 😊

File: examples/offline_inference/encoder_decoder_multimodal.py [hunk 12]
```diff
@@ -121,6 +166,7 @@ def run_whisper():
     "florence2": run_florence2,
     "mllama": run_mllama,
     "whisper": run_whisper,
+    "donut": run_donut,
```
Comment (DarkLight1337) [inline examples/offline_inference/encoder_decoder_multimodal.py]: Please keep the placement in alphabetical order

File: tests/models/multimodal/processing/test_common.py [hunk 13]
```diff
@@ -166,6 +166,7 @@ def _test_processing_correctness(
     "paligemma": False,
     "ultravox": False,
     "whisper": False,
+    "donut": False,
```
Comment (DarkLight1337) [inline tests/models/multimodal/processing/test_common.py]: Please keep the placement in alphabetical order

File: tests/models/multimodal/processing/test_common.py [hunk 14]
```diff
@@ -325,6 +326,7 @@ def _test_processing_correctness_one(
     "omni-research/Tarsier-7b",
     "omni-research/Tarsier2-Recap-7b",
     "mistralai/Voxtral-Mini-3B-2507",
+    "naver-clova-ix/donut-base-finetuned-docvqa",
```
Comment (DarkLight1337) [inline tests/models/multimodal/processing/test_common.py]: Please keep the placement in alphabetical order

File: tests/models/registry.py [hunk 15]
```diff
@@ -522,6 +522,9 @@ def check_available_online(
                                                          trust_remote_code=True),  # noqa: E501
     "MllamaForConditionalGeneration": _HfExamplesInfo("meta-llama/Llama-3.2-11B-Vision-Instruct"),  # noqa: E501
     "WhisperForConditionalGeneration": _HfExamplesInfo("openai/whisper-large-v3"),  # noqa: E501
+    "DonutForConditionalGeneration": _HfExamplesInfo("naver-clova-ix/donut-base-finetuned-docvqa",  # noqa: E501
```
Comment (DarkLight1337) [inline tests/models/registry.py]: Please keep the placement in alphabetical order

**Review by DarkLight1337 — APPROVED** [review]

**Comment by DarkLight1337** [comment]
Thanks!

File: examples/offline_inference/dolphin.py
Comment (hmellor) [reply examples/offline_inference/dolphin.py:1]: It looks like the independent dolphin example still got merged. We don't really want to have model specific examples as that will clutter the examples and make it harder for new users to find what they need
